pkg/sysfs/sst.go

```diff
@@ -0,0 +1,428 @@
+	if info.BFSupported {
+		info.BFCores = IDSet{}
+
+		punitCoreIDs := make(map[ID]IDSet, len(ids))
+		var maxPunitCore ID
+		for _, id := range ids {
+			pc, err := punitCPU(id)
+			if err != nil {
+				return info, err
+			}
+			punitCore := pc >> 1
+			if _, ok := punitCoreIDs[punitCore]; !ok {
+				punitCoreIDs[punitCore] = IDSet{}
+			}
+			punitCoreIDs[punitCore].Add(id)
+			if punitCore > maxPunitCore {
+				maxPunitCore = punitCore
+			}
+		}
+
+		// Read out core masks in batches of 32 (32 bits per response)
+		for i := 0; i <= int(maxPunitCore)/32; i++ {
+			if rsp, err = sendMboxCmd(cpu, CONFIG_TDP, CONFIG_TDP_PBF_GET_CORE_MASK_INFO, uint32(info.PPCurrentLevel+(i<<8))); err != nil {
+				return info, fmt.Errorf("failed to read SST BF core mask (#%d): %v", i, err)
+			}
+			for bit := 0; bit < 32; bit++ {
+				if isBitSet(rsp, uint32(bit)) {
+					info.BFCores.Add(punitCoreIDs[ID(i*32+bit)].Members()...)
+				}
+			}
+		}
+	}
+
+	// Read core-power feature info
+	if rsp, err = sendMboxCmd(cpu, READ_PM_CONFIG, PM_FEATURE, 0); err != nil {
+		return info, fmt.Errorf("failed to read SST CP info: %v", err)
+	}
+
+	info.CPSupported = isBitSet(rsp, 0)
+	info.CPEnabled = isBitSet(rsp, 16)
+
+	if info.CPSupported {
+		if rsp, err = sendMboxCmd(cpu, CONFIG_CLOS, CLOS_PM_QOS_CONFIG, 0); err != nil {
+			return info, fmt.Errorf("failed to read SST CP status: %v", err)
+		}
+
+		closEnabled := isBitSet(rsp, 1)
+		if closEnabled != info.CPEnabled {
+			log.Warn("SST firmware returned conflicting CP enabled status %v vs. %v", info.CPEnabled, closEnabled)
+		}
+		info.CPEnabled = closEnabled
+		info.CPPriority = CPPriorityType(getBits(rsp, 2, 2))
+
+		for i := uint32(0); i < NumClos; i++ {
+			if rsp, err = sendMMIOCmd(cpu, (i<<2)+PM_CLOS_OFFSET); err != nil {
+				return info, fmt.Errorf("failed to read SST CLOS #%d info: %v", i, err)
+			}
+			info.ClosInfo[i] = SstClosInfo{
+				EPP:                  int(getBits(rsp, 0, 3)),
+				ProportionalPriority: int(getBits(rsp, 4, 7)),
+				MinFreq:              int(getBits(rsp, 8, 15)),
+				MaxFreq:              int(getBits(rsp, 16, 23)),
+				DesiredFreq:          int(getBits(rsp, 24, 31)),
+			}
+		}
+	}
+
+	return info, nil
+}
+
+func getCPUClosID(cpu ID) (int, error) {
+	p, err := punitCPU(cpu)
+	if err != nil {
+		return -1, err
+	}
+	punitCore := uint32(p) >> 1
+
+	rsp, err := sendMMIOCmd(cpu, (punitCore<<2)+PQR_ASSOC_OFFSET)
+	if err != nil {
+		return 0, fmt.Errorf("failed to read CLOS number of cpu %d: %v", cpu, err)
+	}
+	return int(getBits(rsp, 16, 17)), nil
+}
+
+func getCPUClosIDs(cpus []ID) ([]int, error) {
+	ret := make([]int, len(cpus))
+	for i, cpu := range cpus {
+		id, err := getCPUClosID(cpu)
+		if err != nil {
+			return ret, err
+		}
+		ret[i] = id
+	}
+	return ret, nil
+}
+
+// IOCTL numbers
+// Derived from kernel headers linux/isst_if.h
+const (
+	ISST_IF_GET_PHY_ID   = 0x000000c008fe01
+	ISST_IF_IO_CMD       = 0x0000004008fe02
+	ISST_IF_MBOX_COMMAND = 0x000000c008fe03
+)
+
```

Comment:
those would be good to move to separate file and use cgo generation, so those constants can be easily updated from linux header files. also, would be good to split sst low level things into separate files and mark them buildable only on linux/x86-64 arch, as ioctl numbers are arch dependant.

Reply:
> those would be good to move to separate file and use cgo generation, so those constants can be easily updated from linux header files. also, would be good to split sst low level things into separate files and mark them buildable only on linux/x86-64 arch, as ioctl numbers are arch dependant.

I originally had the lower and higher level interface in separate files but joined them when moving the code over to `pkg/sysfs`. But I now agree that it's cleaner to keep separate. Changed back to two separate files.

Good point about the ioctl numbers - I didn't realize that they may vary between architectures. I now implemented auto-generation for the definitions and types (structs) derived from the Linux kernel source tree.